### PR TITLE
Added the option to run multitask training for ranking tasks.

### PR DESCRIPTION
**Pull request by zhoutong-fu**
# Description

Detext clients have asked for multitask learning on ranking tasks so that the model can benefit from different subtasks. After reviewing surveys of multitask learning literature, we've decided to add the basic yet very popular hard-parameter sharing neural network structures to Detext framework. 

This implementation assumes that all subtasks share the same deep features while wide features and labels are different, which suggests that both MLP and LTR layers are task-specific and other layers are shared across subtasks. The implementation also assumes individual task losses to be optimized separately and each training record only contains features/labels for one subtask. For training records with more than one sub-tasks, please split them into multiple records.

In summary, this implementation supports multitask training when
* all deep features are shared across all subtasks while wide features are not (task-specific)
* all subtasks are ranking tasks and use the same loss function
* each training record contains features/labels for one of the subtasks

This implementation does **NOT** support
* mixing ranking tasks with classification tasks
* sharing a proportion of deep features
* training record with more than one subtasks (please split them into multiple records)

## How to run multitask training?
* Add `task_id` in the training data as a list of int64. The list should contain only one entry, representing the single subtask for the underlying training record.
* Concatenate wide features of all subtasks into a list and pads 0s for those unrelated to the underlying subtask. For example, suppose task 1 has 5 wide features and task 2 has 10 wide features, then the final wide feature is a list of 15 entries with the first 5 padded with 0s for task2 data records and the last 10 padded with 0s for task 1 data records.
* Follow `run_detext_multitask.py` for other input parameters

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## List all changes 
Please list all changes in the commit.
* Added task_id in `data_fn.py`. task_id is a list of type `int64` and only contains 1 entry (same as query or uid).
* Added task-specific MLP and LTR layers in `deep_match.py`
* Modified `misc_utils.py` to specify and organize required parameters/features
* Updated serving_input_fn, multitask weights in `train.py`
* Added sample data for testing:
  * `multitask_examples.tfrecord` / `test.tfrecord`: sample multitask training data
  * `run_detext_multitask.sh`: sample script to run multitask training
* Added and updated unit tests

# Testing
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

* `pytest` passed
* `bash run_detext.sh` showed consistent results
* `bash run_detext_multitask.sh` partial outputs shown below:

**Trainable variables**
```
INFO:tensorflow:# Trainable variables
INFO:tensorflow:  w_embedding:0, (100, 32), /device:CPU:0
INFO:tensorflow:  cnn/query_cnn_3/kernel:0, (3, 32, 1, 50), /device:CPU:0
INFO:tensorflow:  cnn/query_cnn_3/bias:0, (50,), /device:CPU:0
INFO:tensorflow:  cnn/doc_cnn_0_3/kernel:0, (3, 32, 1, 50), /device:CPU:0
INFO:tensorflow:  cnn/doc_cnn_0_3/bias:0, (50,), /device:CPU:0
INFO:tensorflow:  cnn/doc_cnn_1_3/kernel:0, (3, 32, 1, 50), /device:CPU:0
INFO:tensorflow:  cnn/doc_cnn_1_3/bias:0, (50,), /device:CPU:0
INFO:tensorflow:  wide_ftr_norm_w:0, (10,), /device:CPU:0
INFO:tensorflow:  wide_ftr_norm_b:0, (10,), /device:CPU:0
INFO:tensorflow:  task_0_hidden_projection_0/kernel:0, (12, 100), /device:CPU:0
INFO:tensorflow:  task_0_hidden_projection_0/bias:0, (100,), /device:CPU:0
INFO:tensorflow:  task_0_scoring/kernel:0, (100, 1), /device:CPU:0
INFO:tensorflow:  task_0_scoring/bias:0, (1,), /device:CPU:0
INFO:tensorflow:  task_1_hidden_projection_0/kernel:0, (12, 100), /device:CPU:0
INFO:tensorflow:  task_1_hidden_projection_0/bias:0, (100,), /device:CPU:0
INFO:tensorflow:  task_1_scoring/kernel:0, (100, 1), /device:CPU:0
INFO:tensorflow:  task_1_scoring/bias:0, (1,), /device:CPU:0
INFO:tensorflow:total bert parameters:
INFO:tensorflow:0
```
**Eval results of best model on test data**
```
INFO:tensorflow:global_step = 50
INFO:tensorflow:loss = 0.992338
INFO:tensorflow:metric/ndcg@10 = 0.99576235
```
**Test Configuration**:
* Firmware version:
* Hardware:
* Toolchain:
* SDK:

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/detext/train/data_fn.py [hunk 1]
```diff
@@ -310,7 +310,7 @@ def process_data(record, features_schema):
         features = dict()
         for name in feature_names:
             t = example[name]
-            if name != 'uid':
+            if name not in ('uid', 'task_id'):
```
Comment (StarWang) [inline src/detext/train/data_fn.py:313]: Nice refactoring on this. I don't even know tuple could be passed in

File: src/detext/train/data_fn.py [hunk 2]
```diff
@@ -80,9 +80,9 @@ def _get_tfrecord_feature_parsing_schema(feature_names):
         elif name == 'weight':
             features_tfr[name] = tf.FixedLenFeature(shape=[1], dtype=tf.float32)
             # Default uid as feature for detext integration, will be -1 by default if not present in data
-        elif name == 'uid':
+        elif name in ('uid', 'task_id'):
             features_tfr[name] = tf.FixedLenFeature(shape=[1], dtype=tf.int64)
-        elif name.startswith('doc_') or name.startswith('docId_'):
+        elif name.startswith(('doc_', 'docId_')):
```
Comment (StarWang) [inline src/detext/train/data_fn.py:85]: Nice refactoring on this. I don't even know tuple could be passed in

File: src/detext/train/train.py [hunk 3]
```diff
@@ -298,6 +303,16 @@ def model_fn(features, labels, mode, params):
 
     group_size_field = features['group_size'] if mode != tf.estimator.ModeKeys.PREDICT else None
 
+    # For multitask training
+    task_id_field = features.get('task_id', None)
+
+    # Override the default weight for multitask from input params task_ids and task_weights
+    if params.task_ids is not None:
+        weight = tf.zeros(shape=tf.shape(weight), dtype="float32")
+        for task_id, task_weight in params.task_ids.items():
```
Comment (StarWang) [inline src/detext/train/train.py:312]: Before reading this line, I don't know the item of params.task_ids is task_weight. Can you have a better name for the variable?

File: src/detext/resources/vocab-100.txt [hunk 4]
```diff
@@ -0,0 +1,100 @@
+[PAD]
```
Comment (StarWang) [inline src/detext/resources/vocab-100.txt:1]: Why do we need this file?

File: src/detext/run_detext.py [hunk 5]
```diff
@@ -146,6 +146,12 @@ def add_arguments(parser):
 
     parser.add_argument('--use_horovod', type=str2bool, default=False, help="whether to use horovod for sync distributed training")
 
+    # multitask training related
+    parser.add_argument('--task_ids', type=str, default=None,
+                        help="All types of task IDs for multitask training. E.g., 1,2,3")
+    parser.add_argument('--task_weights', type=str, default=None,
```
Comment (StarWang) [inline src/detext/run_detext.py:152]: Suggest justifying why we need "task_weights" aside from "weight" in the tfrecord in the description. From my point of view, this addition is valid, but I do want to hear about your opinions

File: src/detext/train/data_fn.py (hunk 2 above)
Comment (StarWang) [inline src/detext/train/data_fn.py:85]: Nice refactoring. I don't even know startswith can take in a tuple

File: src/detext/train/train.py [hunk 6]
```diff
@@ -210,6 +210,8 @@ def serving_input_fn(hparams):
         "weight_placeholder", [], tf.float32, 'weight', hparams.feature_names, tf.constant([1.0], dtype=tf.float32))
     label_placeholder, label = train_helper.create_placeholder_for_ftrs(
         "label_placeholder", [None], tf.float32, 'label', hparams.feature_names)
+    task_id_placeholder, task_id = train_helper.create_placeholder_for_ftrs(
```
Comment (StarWang) [inline src/detext/train/train.py:213]: Suggest adding the info of type and shape of task id in the PR context

File: src/detext/train/train.py [hunk 7]
```diff
@@ -298,6 +303,16 @@ def model_fn(features, labels, mode, params):
...
+    # For multitask training
+    task_id_field = features.get('task_id', None)
+
+    # Override the default weight for multitask from input params task_ids and task_weights
+    if params.task_ids is not None:
+        weight = tf.zeros(shape=tf.shape(weight), dtype="float32")
+        for task_id, task_weight in params.task_ids.items():
+            task_mask = tf.cast(tf.equal(task_id_field, task_id), dtype=tf.float32)
+            weight += task_weight * task_mask
+
```
Comment (StarWang) [inline src/detext/train/train.py:315]: weight and task_weight are orthogonal. Multiplying instead of overriding them is more appropriate (consider the case where some samples for a task are downsampled)

File: src/detext/utils/misc_utils.py [hunk 8]
```diff
@@ -156,6 +156,20 @@ def extend_hparams(hparams):
     # L1 and L2 scale must be non-negative values
     assert hparams.l1 is None or hparams.l1 >= 0, "l1 scale must be non-negative"
     assert hparams.l2 is None or hparams.l2 >= 0, "l1 scale must be non-negative"
+
+    # Multi-task training: currently only support ranking tasks with both deep and wide features
+    if hparams.task_ids:
+        # Check related inputs for multi-task training
+        assert 'wide_ftrs_sp_idx' not in hparams.feature_names, "multi-task with sparse features not supported"
```
Comment (StarWang) [inline src/detext/utils/misc_utils.py:163]: Good point

File: src/detext/train/train.py (hunk 3 above)
Comment (zhoutong-linkedin) [reply src/detext/train/train.py:312]: Thanks for the suggestion. I'll update with a better name and avoid overriding the input param `task_ids`.

File: src/detext/run_detext.py (hunk 5 above)
Comment (zhoutong-linkedin) [reply src/detext/run_detext.py:152]: This is a good question. I checked with detext users about this parameter. The weight feature is also valid for multitask training but it'll be more convenient to specify them as configurable parameters than features so that users do not need to regenerate training data when tuning the weight for each task.

File: src/detext/train/train.py (hunk 7 above)
Comment (zhoutong-linkedin) [reply src/detext/train/train.py:315]: It makes sense to separate them conceptually. In practice, they may not be differentiable. For example, the sample weights are equivalent to the task weights when the loss is calculated as the weighted avg of different records/tasks. But I agree it'll be flexible to separate them here for flexibility. I'll update it.

File: src/detext/resources/vocab-100.txt (hunk 4 above)
Comment (zhoutong-linkedin) [reply src/detext/resources/vocab-100.txt:1]: The other vocab file is a bit too small. For local testing, I used this vocab with 100 tokens to run on the sample data.

File: src/detext/train/train.py (hunk 6 above)
Comment (zhoutong-linkedin) [reply src/detext/train/train.py:213]: Thanks, I will do that. Also, reply to your general comments (which I was not able to reply directly) on the PR context, I'll update more details for what this implementation can vs cannot do. I'm also planning to update related documentation after resolving comments/questions from other reviewers.

File: src/detext/resources/vocab-100.txt (hunk 4 above)
Comment (zhoutong-linkedin) [reply src/detext/resources/vocab-100.txt:1]: removed this one in the latest update.

File: src/detext/train/train.py (hunk 3 above)
Comment (zhoutong-linkedin) [reply src/detext/train/train.py:312]: Added a new param `task_id_weight_map` that zips task_id and weight into a map.

**Comment by xwli-chelsea** [comment]
Wide_ftrs also need to match among tasks right?

**Comment by zhoutong-linkedin** [comment]
> Comments left. Thanks for the implementation for multitask training!
> 
> Let's add more details on what this PR is not implementing here. I.e., although different losses are computed, these losses are using the same loss_fn. Also, we do not support mixing ranking tasks with classification tasks.

Updated in the PR description. Let me know if it looks good to youl

**Comment by zhoutong-linkedin** [comment]
> Wide_ftrs also need to match among tasks right?

Updated the PR and added a section on how to run multitask training.

File: src/detext/train/train.py [hunk 9]
```diff
@@ -298,6 +303,17 @@ def model_fn(features, labels, mode, params):
...
 
+    # For multitask training
+    task_id_field = features.get('task_id', None)
+
+    # Update the weight with each task's weight such that weight per document = weight * task_weight
+    if params.task_id_weight_map is not None:
+        weight_adj = tf.zeros(shape=tf.shape(weight), dtype="float32")
+        for task_id, task_weight in params.task_id_weight_map.items():
+            task_mask = tf.cast(tf.equal(task_id_field, task_id), dtype=tf.float32)
+            weight_adj += task_weight * task_mask
```
Comment (StarWang) [inline src/detext/train/train.py:314]: This `for ... ` loop makes me a bit confused and what does adj mean here? adjusted? Just to confirm, the task_weight * task_mask will be non zero once in the loop right?

In fact you can use a implementation such as
task_weights = params.task_weights
task_ids = params.task_ids
task_mask = tf.cast(tf.equal(task_weight, task_id)...
weight_adj = tf.reduce_sum(task_mask*task_weight)

This is easier to read and more of a vectorized programming style

Comment (zhoutong-linkedin) [reply src/detext/train/train.py:314]: To avoid the loop, I will use tf.broadcast to expand task_id_field into a shape of [batch_size, num_tasks] and then perform similar ops and tf.reduce_sum. Just want to clarify that both task_weight and task_id are lists, e.g. [0, 1, 2] for task_id, [0.1, 0.3, 0.6] for task_weight, where task_id_field is a tensor of shape [batch_size, ].

File: src/detext/train/train.py [hunk 10]
```diff
@@ -298,6 +303,18 @@ def model_fn(features, labels, mode, params):
...
+    # For multitask training
+    task_id_field = features.get('task_id', None)  # shape=[batch_size,]
+
+    # Update the weight with each task's weight such that weight per document = weight * task_weight
+    if params.task_ids is not None:
+        task_ids = params.task_ids  # e.g. [0, 1, 2]
+        task_weights = params.task_weights  # e.g. [0.1, 0.3, 0.6]
+        # Expand task_id_field with shape [batch_size, num_tasks]
+        expanded_task_id_field = tf.transpose(tf.broadcast_to(task_id_field, [len(task_ids), tf.shape(task_id_field)[0]]))
+        task_mask = tf.cast(tf.equal(expanded_task_id_field, task_ids), dtype=tf.float32)
```
Comment (StarWang) [inline src/detext/train/train.py:315]: Just make sure that results are correct here with the implicit broadcasting of task_ids and task_weights

Comment (zhoutong-linkedin) [reply src/detext/train/train.py:315]: Thanks. I tested locally this part with sample data. Did not add the unit test as it's more about tf.broadcast_to practice.
